File: src/test/java/com/zepben/vertxutils/json/JsonUtilsTest.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.json;

import com.zepben.annotations.EverythingIsNonnullByDefault;
import com.zepben.testutils.junit.SystemLogExtension;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.RegisterExtension;

import javax.annotation.Nullable;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collector;
import java.util.stream.Stream;

import static com.zepben.testutils.exception.ExpectException.expect;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.contains;
import static org.hamcrest.Matchers.nullValue;
import static org.hamcrest.core.IsEqual.equalTo;

@EverythingIsNonnullByDefault
public class JsonUtilsTest {

    @RegisterExtension
    static SystemLogExtension systemOut = SystemLogExtension.SYSTEM_OUT.captureLog().muteOnSuccess();

    @RegisterExtension
    static SystemLogExtension systemErr = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess();

    private static final String VALID_KEY = "key";
    private static final String MISSING_KEY = "key2";

    private final int integerVal1 = 111;
    private final int integerVal2 = 222;
    private final int integerVal3 = 333;
    private final double doubleVal1 = 1.11;
    private final double doubleVal2 = 2.22;
    private final String stringVal = "value";
    private final Path pathVal = Paths.get("valid/path");
    private final String illegalPath = "illegal\0path";

    private final JsonObject jsonObjectVal1 = createObjectWithValue(stringVal);
    private final JsonObject jsonObjectVal2 = createObjectWithValue(doubleVal2);
    private final JsonObject jsonObjectVal3 = createObjectWithValue(illegalPath);
    private final JsonArray jsonArrayOfJsonObjects = createArray(jsonObjectVal1, jsonObjectVal2);
    private final JsonArray jsonArrayOfIntegers = createArray(integerVal1, integerVal2, integerVal3);
    private final JsonArray jsonArrayOfMixed = createArray(jsonObjectVal1, integerVal2);

    @Test
    public void extractValue() {
        validateExtractor(JsonUtils.INSTANCE::extractOptionalValue, JsonUtils.INSTANCE::extractRequiredValue, jsonObjectVal1);
        validateExtractor(JsonUtils.INSTANCE::extractOptionalValue, JsonUtils.INSTANCE::extractRequiredValue, integerVal1);
        validateExtractor(JsonUtils.INSTANCE::extractOptionalValue, JsonUtils.INSTANCE::extractRequiredValue, stringVal);
    }

    @Test
    public void extractObject() {
        validateExtractor(JsonUtils.INSTANCE::extractOptionalObject, JsonUtils.INSTANCE::extractRequiredObject, jsonObjectVal1, stringVal, "object");
    }

    @Test
    public void extractArray() {
        validateExtractor(JsonUtils.INSTANCE::extractOptionalArray, JsonUtils.INSTANCE::extractRequiredArray, jsonArrayOfJsonObjects, stringVal, "array");
        validateExtractor(JsonUtils.INSTANCE::extractOptionalArray, JsonUtils.INSTANCE::extractRequiredArray, jsonArrayOfIntegers, stringVal, "array");
    }

    @Test
    public void extractString() {
        validateExtractor(JsonUtils.INSTANCE::extractOptionalString, JsonUtils.INSTANCE::extractRequiredString, stringVal, integerVal1, "string");
    }

    @Test
    public void extractInt() {
        validateExtractor(JsonUtils.INSTANCE::extractOptionalInt, JsonUtils.INSTANCE::extractRequiredInt, integerVal1, stringVal, "integer");
    }

    @Test
    public void extractDouble() {
        validateExtractor(JsonUtils.INSTANCE::extractOptionalDouble, JsonUtils.INSTANCE::extractRequiredDouble, doubleVal1, stringVal, "double");
        validateExtractor(JsonUtils.INSTANCE::extractOptionalDouble, JsonUtils.INSTANCE::extractRequiredDouble, Double.NaN, stringVal, "double");
    }

    @Test
    public void extractPath() {
        validateExtractor(JsonUtils.INSTANCE::extractOptionalPath, JsonUtils.INSTANCE::extractRequiredPath, pathVal, Path::toString, doubleVal2, "path");
        validateExtractor(JsonUtils.INSTANCE::extractOptionalPath, JsonUtils.INSTANCE::extractRequiredPath, pathVal, Path::toString, illegalPath, "path");
    }

    @Test
    public void extractObjectList() {
        List<JsonObject> jsonObjects = Arrays.asList(jsonObjectVal1, jsonObjectVal2);

        validateExtractor(JsonUtils.INSTANCE::extractOptionalObjectList,
            JsonUtils.INSTANCE::extractRequiredObjectList,
            jsonObjects,
            JsonArray::new,
            Arrays.asList(integerVal1, integerVal2),
            "list of objects");

        validateExtractor(JsonUtils.INSTANCE::extractOptionalObjectList,
            JsonUtils.INSTANCE::extractRequiredObjectList,
            jsonObjects,
            JsonArray::new,
            stringVal,
            "array");
    }

    @Test
    public void extractStringList() {
        List<String> strings = Arrays.asList(stringVal, stringVal);

        validateExtractor(JsonUtils.INSTANCE::extractOptionalStringList,
            JsonUtils.INSTANCE::extractRequiredStringList,
            strings,
            JsonArray::new,
            Arrays.asList(integerVal1, integerVal2),
            "list of strings");

        validateExtractor(JsonUtils.INSTANCE::extractOptionalStringList,
            JsonUtils.INSTANCE::extractRequiredStringList,
            strings,
            JsonArray::new,
            stringVal,
            "array");
    }

    @Test
    public void extractIntList() {
        List<Integer> integers = Arrays.asList(integerVal1, integerVal2);

        validateExtractor(JsonUtils.INSTANCE::extractOptionalIntList,
            JsonUtils.INSTANCE::extractRequiredIntList,
            integers,
            JsonArray::new,
            createArray(stringVal, stringVal),
            "list of integers");

        validateExtractor(JsonUtils.INSTANCE::extractOptionalIntList,
            JsonUtils.INSTANCE::extractRequiredIntList,
            integers,
            JsonArray::new,
            stringVal,
            "array");
    }

    @Test
    public void extractDoubleList() {
        List<Double> doubles = Arrays.asList(doubleVal1, doubleVal2, Double.NaN);

        validateExtractor(JsonUtils.INSTANCE::extractOptionalDoubleList,
            JsonUtils.INSTANCE::extractRequiredDoubleList,
            doubles,
            JsonArray::new,
            createArray(stringVal, stringVal),
            "list of doubles");

        validateExtractor(JsonUtils.INSTANCE::extractOptionalDoubleList,
            JsonUtils.INSTANCE::extractRequiredDoubleList,
            doubles,
            JsonArray::new,
            stringVal,
            "array");
    }

    @Test
    public void extractObjectListOfList() {
        List<List<JsonObject>> lists = Arrays.asList(Arrays.asList(jsonObjectVal1, jsonObjectVal2), Collections.singletonList(jsonObjectVal3));

        Function<List<List<JsonObject>>, Object> listsToJsonArray = l -> l
            .stream()
            .map(JsonArray::new)
            .collect(Collector.of(JsonArray::new, JsonArray::add, JsonArray::addAll));

        validateExtractor(JsonUtils.INSTANCE::extractOptionalObjectListOfList,
            JsonUtils.INSTANCE::extractRequiredObjectListOfList,
            lists,
            listsToJsonArray,
            createArray(stringVal, stringVal),
            "list of object lists");

        validateExtractor(JsonUtils.INSTANCE::extractOptionalObjectListOfList,
            JsonUtils.INSTANCE::extractRequiredObjectListOfList,
            lists,
            listsToJsonArray,
            createArray(jsonArrayOfIntegers),
            "list of object lists");
    }

    @Test
    public void convertsJsonArrayToObjectList() {
        assertThat(JsonUtils.INSTANCE.convertToObjectList(jsonArrayOfJsonObjects), contains(jsonObjectVal1, jsonObjectVal2));

        expect(() -> JsonUtils.INSTANCE.convertToObjectList(jsonArrayOfIntegers))
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage("JSON array is not a collection of expected types.");

        expect(() -> JsonUtils.INSTANCE.convertToObjectList(jsonArrayOfMixed))
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage("JSON array is not a collection of expected types.");
    }

    @Test
    public void convertsJsonArrayToList() {
        assertThat(JsonUtils.INSTANCE.convertToList(jsonArrayOfJsonObjects, JsonArray::getJsonObject), contains(jsonObjectVal1, jsonObjectVal2));
        assertThat(JsonUtils.INSTANCE.convertToList(jsonArrayOfIntegers, JsonArray::getInteger), contains(integerVal1, integerVal2, integerVal3));
        assertThat(JsonUtils.INSTANCE.convertToList(jsonArrayOfMixed, JsonArray::getValue), contains(jsonObjectVal1, integerVal2));

        expect(() -> JsonUtils.INSTANCE.convertToList(jsonArrayOfJsonObjects, JsonArray::getInteger))
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage("JSON array is not a collection of expected types.");

        expect(() -> JsonUtils.INSTANCE.convertToList(jsonArrayOfIntegers, JsonArray::getJsonObject))
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage("JSON array is not a collection of expected types.");

        expect(() -> JsonUtils.INSTANCE.convertToList(jsonArrayOfMixed, JsonArray::getDouble))
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage("JSON array is not a collection of expected types.");
    }

    @Test
    public void convertsJsonArrayToListWithExpectedCount() {
        assertThat(JsonUtils.INSTANCE.convertToList(jsonArrayOfJsonObjects, JsonArray::getJsonObject, jsonArrayOfJsonObjects.size()), contains(jsonObjectVal1, jsonObjectVal2));
        assertThat(JsonUtils.INSTANCE.convertToList(jsonArrayOfIntegers, JsonArray::getInteger, jsonArrayOfIntegers.size()), contains(integerVal1, integerVal2, integerVal3));
        assertThat(JsonUtils.INSTANCE.convertToList(jsonArrayOfMixed, JsonArray::getValue, jsonArrayOfMixed.size()), contains(jsonObjectVal1, integerVal2));

        expect(() -> JsonUtils.INSTANCE.convertToList(jsonArrayOfJsonObjects, JsonArray::getJsonObject, 3))
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage("Invalid number of records in list. Expected exactly 3, found 2.");

        expect(() -> JsonUtils.INSTANCE.convertToList(jsonArrayOfIntegers, JsonArray::getJsonObject, 2))
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage("Invalid number of records in list. Expected exactly 2, found 3.");

        expect(() -> JsonUtils.INSTANCE.convertToList(jsonArrayOfMixed, JsonArray::getInteger))
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage("JSON array is not a collection of expected types.");
    }

    private JsonObject createObjectWithValue(@Nullable Object value) {
        JsonObject json = new JsonObject();
        if (value != null)
            json.put(VALID_KEY, value);
        else
            json.putNull(VALID_KEY);
        return new JsonObject(json.encode());
    }

    private JsonArray createArray(@Nullable Object... values) {
        JsonArray json = new JsonArray();
        if (values != null)
            Stream.of(values).forEach(json::add);
        else
            json.addNull();
        return new JsonArray(json.encode());
    }

    private <T> void validateExtractor(BiFunction<JsonObject, String, T> optionalValueExtractor,
                                       BiFunction<JsonObject, String, T> requiredValueExtractor,
                                       T expectedValue) {
        validateExtractor(optionalValueExtractor, requiredValueExtractor, expectedValue, value -> value);
    }

    private <T> void validateExtractor(BiFunction<JsonObject, String, T> optionalValueExtractor,
                                       BiFunction<JsonObject, String, T> requiredValueExtractor,
                                       T expectedValue,
                                       Function<T, Object> valueConverter) {
        JsonObject validObject = createObjectWithValue(valueConverter.apply(expectedValue));

        assertThat(optionalValueExtractor.apply(validObject, VALID_KEY), equalTo(expectedValue));
        assertThat(optionalValueExtractor.apply(validObject, MISSING_KEY), nullValue());

        assertThat(requiredValueExtractor.apply(validObject, VALID_KEY), equalTo(expectedValue));

        expect(() -> {@SuppressWarnings("unused") var unused = requiredValueExtractor.apply(validObject, MISSING_KEY);})
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage("No value found for required key 'key2'.");

    }

    private <T> void validateExtractor(BiFunction<JsonObject, String, T> optionalValueExtractor,
                                       BiFunction<JsonObject, String, T> requiredValueExtractor,
                                       T expectedValue,
                                       Object invalidValue,
                                       String description) {
        validateExtractor(optionalValueExtractor, requiredValueExtractor, expectedValue, value -> value, invalidValue, description);
    }

    private <T> void validateExtractor(BiFunction<JsonObject, String, T> optionalValueExtractor,
                                       BiFunction<JsonObject, String, T> requiredValueExtractor,
                                       T expectedValue,
                                       Function<T, Object> valueConverter,
                                       Object invalidValue,
                                       String description) {
        validateExtractor(optionalValueExtractor, requiredValueExtractor, expectedValue, valueConverter);

        JsonObject invalidObject = createObjectWithValue(invalidValue);

        expect(() -> {@SuppressWarnings("unused") var unused = optionalValueExtractor.apply(invalidObject, VALID_KEY);})
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage(String.format("Value for 'key' is not a valid %s.", description));

        expect(() -> {@SuppressWarnings("unused") var unused = requiredValueExtractor.apply(invalidObject, VALID_KEY);})
            .toThrow(JsonUtils.ParsingException.class)
            .withMessage(String.format("Value for 'key' is not a valid %s.", description));
    }

}

File: src/test/java/com/zepben/vertxutils/json/JsonUtilsTest.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.json

import com.zepben.testutils.exception.ExpectException.Companion.expect
import com.zepben.testutils.junit.SystemLogExtension
import com.zepben.vertxutils.json.JsonUtils.convertToList
import com.zepben.vertxutils.json.JsonUtils.convertToObjectList
import com.zepben.vertxutils.json.JsonUtils.extractOptionalArray
import com.zepben.vertxutils.json.JsonUtils.extractOptionalDouble
import com.zepben.vertxutils.json.JsonUtils.extractOptionalDoubleList
import com.zepben.vertxutils.json.JsonUtils.extractOptionalInt
import com.zepben.vertxutils.json.JsonUtils.extractOptionalIntList
import com.zepben.vertxutils.json.JsonUtils.extractOptionalObject
import com.zepben.vertxutils.json.JsonUtils.extractOptionalObjectList
import com.zepben.vertxutils.json.JsonUtils.extractOptionalObjectListOfList
import com.zepben.vertxutils.json.JsonUtils.extractOptionalPath
import com.zepben.vertxutils.json.JsonUtils.extractOptionalString
import com.zepben.vertxutils.json.JsonUtils.extractOptionalStringList
import com.zepben.vertxutils.json.JsonUtils.extractOptionalValue
import com.zepben.vertxutils.json.JsonUtils.extractRequiredArray
import com.zepben.vertxutils.json.JsonUtils.extractRequiredDouble
import com.zepben.vertxutils.json.JsonUtils.extractRequiredDoubleList
import com.zepben.vertxutils.json.JsonUtils.extractRequiredInt
import com.zepben.vertxutils.json.JsonUtils.extractRequiredIntList
import com.zepben.vertxutils.json.JsonUtils.extractRequiredObject
import com.zepben.vertxutils.json.JsonUtils.extractRequiredObjectList
import com.zepben.vertxutils.json.JsonUtils.extractRequiredObjectListOfList
import com.zepben.vertxutils.json.JsonUtils.extractRequiredPath
import com.zepben.vertxutils.json.JsonUtils.extractRequiredString
import com.zepben.vertxutils.json.JsonUtils.extractRequiredStringList
import com.zepben.vertxutils.json.JsonUtils.extractRequiredValue
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import io.vertx.kotlin.core.json.array
import io.vertx.kotlin.core.json.json
import io.vertx.kotlin.core.json.obj
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.RegisterExtension
import java.nio.file.Path
import java.nio.file.Paths

class JsonUtilsTest {

    companion object {

        @JvmField
        @RegisterExtension
        val systemOut: SystemLogExtension = SystemLogExtension.SYSTEM_OUT.captureLog().muteOnSuccess()

        private const val VALID_KEY = "key"
        private const val MISSING_KEY = "key2"

    }

    private val integerVal1 = 111
    private val integerVal2 = 222
    private val integerVal3 = 333
    private val doubleVal1 = 1.11
    private val doubleVal2 = 2.22
    private val stringVal = "value"
    private val pathVal: Path = Paths.get("valid/path")
    private val illegalPath = "illegal\u0000path"

    private val jsonObjectVal1: JsonObject = createObjectWithValue(stringVal)
    private val jsonObjectVal2: JsonObject = createObjectWithValue(doubleVal2)
    private val jsonObjectVal3: JsonObject = createObjectWithValue(illegalPath)
    private val jsonArrayOfJsonObjects: JsonArray = createArray(jsonObjectVal1, jsonObjectVal2)
    private val jsonArrayOfIntegers: JsonArray = createArray(integerVal1, integerVal2, integerVal3)
    private val jsonArrayOfMixed: JsonArray = createArray(jsonObjectVal1, integerVal2)

    @Test
    fun extractValue() {
        validateExtractor(
            { obj, json -> obj.extractOptionalValue(json) },
            { obj, json -> obj.extractRequiredValue(json) },
            jsonObjectVal1,
        )
        validateExtractor(
            { obj, json -> obj.extractOptionalValue(json) },
            { obj, json -> obj.extractRequiredValue(json) },
            integerVal1,
        )
        validateExtractor(
            { obj, json -> obj.extractOptionalValue(json) },
            { obj, json -> obj.extractRequiredValue(json) },
            stringVal,
        )
    }

    @Test
    fun extractObject() {
        validateExtractor(
            { obj, json -> obj.extractOptionalObject(json) },
            { obj, json -> obj.extractRequiredObject(json) },
            jsonObjectVal1,
            stringVal,
            "object",
        )
    }

    @Test
    fun extractArray() {
        validateExtractor(
            { obj, json -> obj.extractOptionalArray(json) },
            { obj, json -> obj.extractRequiredArray(json) },
            jsonArrayOfJsonObjects,
            stringVal,
            "array",
        )
        validateExtractor(
            { obj, json -> obj.extractOptionalArray(json) },
            { obj, json -> obj.extractRequiredArray(json) },
            jsonArrayOfIntegers,
            stringVal,
            "array",
        )
    }

    @Test
    fun extractString() {
        validateExtractor(
            { obj, json -> obj.extractOptionalString(json) },
            { obj, json -> obj.extractRequiredString(json) },
            stringVal,
            integerVal1,
            "string",
        )
    }

    @Test
    fun extractInt() {
        validateExtractor(
            { obj, json -> obj.extractOptionalInt(json) },
            { obj, json -> obj.extractRequiredInt(json) },
            integerVal1,
            stringVal,
            "integer",
        )
    }

    @Test
    fun extractDouble() {
        validateExtractor(
            { obj, json -> obj.extractOptionalDouble(json) },
            { obj, json -> obj.extractRequiredDouble(json) },
            doubleVal1,
            stringVal,
            "double",
        )
        validateExtractor(
            { obj, json -> obj.extractOptionalDouble(json) },
            { obj, json -> obj.extractRequiredDouble(json) },
            Double.NaN,
            stringVal,
            "double",
        )
    }

    @Test
    fun extractPath() {
        validateExtractor(
            { obj, json -> obj.extractOptionalPath(json) },
            { obj, json -> obj.extractRequiredPath(json) },
            pathVal,
            doubleVal2,
            "path",
        ) { it.toString() }
        validateExtractor(
            { obj, json -> obj.extractOptionalPath(json) },
            { obj, json -> obj.extractRequiredPath(json) },
            pathVal,
            illegalPath,
            "path",
        ) { it.toString() }
    }

    @Test
    fun extractObjectList() {
        val jsonObjects = listOf(jsonObjectVal1, jsonObjectVal2)

        validateExtractor(
            { obj, json -> obj.extractOptionalObjectList(json) },
            { obj, json -> obj.extractRequiredObjectList(json) },
            jsonObjects,
            listOf(integerVal1, integerVal2),
            "list of objects",
        ) { list -> JsonArray(list) }

        validateExtractor(
            { obj, json -> obj.extractOptionalObjectList(json) },
            { obj, json -> obj.extractRequiredObjectList(json) },
            jsonObjects,
            stringVal,
            "array",
        ) { list -> JsonArray(list) }
    }

    @Test
    fun extractStringList() {
        val strings = listOf(stringVal, stringVal)

        validateExtractor(
            { obj, json -> obj.extractOptionalStringList(json) },
            { obj, json -> obj.extractRequiredStringList(json) },
            strings,
            listOf(integerVal1, integerVal2),
            "list of strings",
        ) { list -> JsonArray(list) }

        validateExtractor(
            { obj, json -> obj.extractOptionalStringList(json) },
            { obj, json -> obj.extractRequiredStringList(json) },
            strings,
            stringVal,
            "array",
        ) { list -> JsonArray(list) }
    }

    @Test
    fun extractIntList() {
        val integers = listOf(integerVal1, integerVal2)

        validateExtractor(
            { obj, json -> obj.extractOptionalIntList(json) },
            { obj, json -> obj.extractRequiredIntList(json) },
            integers,
            createArray(stringVal, stringVal),
            "list of integers",
        ) { list -> JsonArray(list) }

        validateExtractor(
            { obj, json -> obj.extractOptionalIntList(json) },
            { obj, json -> obj.extractRequiredIntList(json) },
            integers,
            stringVal,
            "array",
        ) { list -> JsonArray(list) }
    }

    @Test
    fun extractDoubleList() {
        val doubles = listOf(doubleVal1, doubleVal2, Double.NaN)

        validateExtractor(
            { obj, json -> obj.extractOptionalDoubleList(json) },
            { obj, json -> obj.extractRequiredDoubleList(json) },
            doubles,
            createArray(stringVal, stringVal),
            "list of doubles",
        ) { list -> JsonArray(list) }

        validateExtractor(
            { obj, json -> obj.extractOptionalDoubleList(json) },
            { obj, json -> obj.extractRequiredDoubleList(json) },
            doubles,
            stringVal,
            "array",
        ) { list -> JsonArray(list) }
    }

    @Test
    fun extractObjectListOfList() {
        val lists = listOf(listOf(jsonObjectVal1, jsonObjectVal2), listOf(jsonObjectVal3))

        validateExtractor(
            { obj, json -> obj.extractOptionalObjectListOfList(json) },
            { obj, json -> obj.extractRequiredObjectListOfList(json) },
            lists,
            createArray(stringVal, stringVal),
            "list of object lists",
        ) { list -> JsonArray(list!!.map { JsonArray(it) }) }

        validateExtractor(
            { obj, json -> obj.extractOptionalObjectListOfList(json) },
            { obj, json -> obj.extractRequiredObjectListOfList(json) },
            lists,
            createArray(jsonArrayOfIntegers),
            "list of object lists",
        ) { list -> JsonArray(list!!.map { JsonArray(it) }) }
    }

    @Test
    fun convertsJsonArrayToObjectList() {
        assertThat(
            jsonArrayOfJsonObjects.convertToObjectList(),
            contains(jsonObjectVal1, jsonObjectVal2),
        )

        expect { jsonArrayOfIntegers.convertToObjectList() }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("JSON array is not a collection of expected types.")

        expect { jsonArrayOfMixed.convertToObjectList() }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("JSON array is not a collection of expected types.")
    }

    @Test
    fun convertsJsonArrayToList() {
        assertThat(
            jsonArrayOfJsonObjects.convertToList { obj, pos -> obj.getJsonObject(pos) },
            contains(jsonObjectVal1, jsonObjectVal2),
        )
        assertThat(
            jsonArrayOfIntegers.convertToList { obj, pos -> obj.getInteger(pos) },
            contains(integerVal1, integerVal2, integerVal3),
        )
        assertThat(
            jsonArrayOfMixed.convertToList { obj, pos -> obj.getValue(pos) },
            contains(jsonObjectVal1, integerVal2),
        )

        expect { jsonArrayOfJsonObjects.convertToList { obj, pos -> obj.getInteger(pos) } }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("JSON array is not a collection of expected types.")

        expect { jsonArrayOfIntegers.convertToList { obj, pos -> obj.getJsonObject(pos) } }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("JSON array is not a collection of expected types.")

        expect { jsonArrayOfMixed.convertToList { obj, pos -> obj.getDouble(pos) } }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("JSON array is not a collection of expected types.")
    }

    @Test
    fun convertsJsonArrayToListWithExpectedCount() {
        assertThat(
            jsonArrayOfJsonObjects.convertToList(
                jsonArrayOfJsonObjects.size(),
            ) { arr, pos -> arr.getJsonObject(pos) },
            contains(jsonObjectVal1, jsonObjectVal2),
        )
        assertThat(
            jsonArrayOfIntegers.convertToList(jsonArrayOfIntegers.size()) { obj, pos -> obj.getInteger(pos) },
            contains(integerVal1, integerVal2, integerVal3),
        )
        assertThat(
            jsonArrayOfMixed.convertToList(jsonArrayOfMixed.size()) { obj, pos -> obj.getValue(pos) },
            contains(jsonObjectVal1, integerVal2),
        )

        expect { jsonArrayOfJsonObjects.convertToList(3) { obj, pos -> obj.getJsonObject(pos) } }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("Invalid number of records in list. Expected exactly 3, found 2.")

        expect { jsonArrayOfIntegers.convertToList(2) { obj, pos -> obj.getJsonObject(pos) } }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("Invalid number of records in list. Expected exactly 2, found 3.")

        expect { jsonArrayOfMixed.convertToList { obj, pos -> obj.getInteger(pos) } }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("JSON array is not a collection of expected types.")
    }

    private fun createObjectWithValue(value: Any?): JsonObject =
        json {
            obj(VALID_KEY to value)
        }

    private fun createArray(vararg values: Any?): JsonArray =
        json {
            array(*values)
        }

    private fun <T> validateExtractor(
        optionalValueExtractor: (JsonObject, String) -> T?,
        requiredValueExtractor: (JsonObject, String) -> T,
        expectedValue: T?,
    ) {
        validateExtractor(optionalValueExtractor, requiredValueExtractor, expectedValue) { it }
    }

    private fun <T> validateExtractor(
        optionalValueExtractor: (JsonObject, String) -> T?,
        requiredValueExtractor: (JsonObject, String) -> T,
        expectedValue: T?,
        valueConverter: (T?) -> Any?,
    ) {
        val validObject = createObjectWithValue(valueConverter(expectedValue))

        assertThat(optionalValueExtractor(validObject, VALID_KEY), equalTo(expectedValue))
        assertThat(optionalValueExtractor(validObject, MISSING_KEY), nullValue())

        assertThat(requiredValueExtractor(validObject, VALID_KEY), equalTo(expectedValue))

        expect { requiredValueExtractor(validObject, MISSING_KEY) }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("No value found for required key 'key2'.")
    }

    private fun <T> validateExtractor(
        optionalValueExtractor: (JsonObject, String) -> T?,
        requiredValueExtractor: (JsonObject, String) -> T,
        expectedValue: T?,
        invalidValue: Any,
        description: String,
    ) {
        validateExtractor(optionalValueExtractor, requiredValueExtractor, expectedValue, invalidValue, description) { it }
    }

    private fun <T> validateExtractor(
        optionalValueExtractor: (JsonObject, String) -> T?,
        requiredValueExtractor: (JsonObject, String) -> T,
        expectedValue: T?,
        invalidValue: Any,
        description: String,
        valueConverter: (T?) -> Any?,
    ) {
        validateExtractor(optionalValueExtractor, requiredValueExtractor, expectedValue, valueConverter)

        val invalidObject = createObjectWithValue(invalidValue)

        expect { optionalValueExtractor(invalidObject, VALID_KEY) }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("Value for 'key' is not a valid $description.")

        expect { requiredValueExtractor(invalidObject, VALID_KEY) }
            .toThrow<JsonUtils.ParsingException>()
            .withMessage("Value for 'key' is not a valid $description.")
    }

}
